src/compiler/checker.ts

```diff
@@ -8016,6 +8154,65 @@ module ts {
             }
         }
 
+        function getAwaitedType(type: Type, fallbackType?: Type, seen: boolean[] = []): Type {

```

Comment:
Ah I see. For getAwaitedType, you say you recursively unwrap the promised type. But how can the promised type be recursively wrapped in the first place, given the way that promises resolve?